### PR TITLE
Setup Window prototype in vm context

### PR DESCRIPTION
**Pull request by legendecas**
Refs: https://github.com/nodejs/node/issues/54436#issuecomment-2296539259

At the moment, the global object prototype in the vm context is as follows:

```
Outer Context    |    Vm Context
Object           |    Object
  ^              |      ^
  | prototype    |      | prototype
EventTarget      |    Global
  ^              |      ^
  | prototype    |      | prototype
Window           |      |
  ^              |      |
  | prototype    |      |
window (sandbox) <--- globalThis
```

This PR corrects the prototype in respect to the WebIDL definition as:

```
Outer Context    |    Vm Context
Object           |    Object
  ^              |   
  | prototype    |
EventTarget      |
  ^              |
  | prototype    |
Window (and WindowProperties)
  ^              |      ^
  | prototype    |      | prototype
window (sandbox) <--- globalThis
```

### CODE REVIEW

**Comment by domenic** [comment]
Well, this does look promising! We can try doing a release and seeing if anyone's code breaks, but our test suite is pretty exhaustive, so I'm hopeful.

File: lib/jsdom/browser/Window.js [hunk 1]
```diff
@@ -148,6 +156,10 @@ function setupWindow(windowInstance, { runScripts }) {
 
   windowConstructor.prototype = windowPrototype;
   Object.setPrototypeOf(windowInstance, windowPrototype);
+  if (runScripts === "outside-only" || runScripts === "dangerously") {
+    const global = vm.runInContext("this", windowInstance);
+    Object.setPrototypeOf(global, windowPrototype);
```
Comment (domenic) [inline lib/jsdom/browser/Window.js:161]: Note to self: we could try also setting `EventTarget.prototype.__proto__` to the `Object.prototype` from the vm context. Might be best to do with a major version bump...

**Comment by legendecas** [comment]
The test fails because html/browsers/the-window-object.constructor expect `window.constructor` is not an own property. But in Node.js versions before v22.5.0, all prototype properties are intercepted as own properties of the global object. This is not fixable in userland unless with some strange polyfills.

Is there a way to specify a test only passes on certion versions?

File: lib/jsdom/browser/Window.js (hunk 1 above)
Comment (legendecas) [reply lib/jsdom/browser/Window.js:161]: Yes, I assume this could be a big change so I'd leave it from this PR...

**Comment by asamuzaK** [comment]
> Is there a way to specify a test only passes on certion versions?

Maybe some kind of a reason like [this](https://github.com/jsdom/jsdom/pull/3757#issuecomment-2295545330) in [run-wpts.js](https://github.com/jsdom/jsdom/blob/main/test/web-platform-tests/run-wpts.js#L25)?

**Comment by legendecas** [comment]
> > Is there a way to specify a test only passes on certion versions?
> 
> Maybe some kind of a reason like [this](https://github.com/jsdom/jsdom/pull/3757#issuecomment-2295545330) in [run-wpts.js](https://github.com/jsdom/jsdom/blob/main/test/web-platform-tests/run-wpts.js#L25) and [utils.js](https://github.com/jsdom/jsdom/blob/main/test/web-platform-tests/utils.js#L12)?

Yes, in v22.5.0 the `html/browsers/the-window-object.constructor` passes because `Object.hasOwn` behaves correctly on the global object.

**Comment by asamuzaK** [comment]
Regarding https://github.com/jsdom/jsdom/issues/3688

The test case below fails `test2` and `test3`:

``` html
<!DOCTYPE html>
<meta charset="utf-8">
<title>Access window properties</title>
<link rel="help" href="https://github.com/jsdom/jsdom/issues/3688">

<script src="/resources/testharness.js"></script>
<script src="/resources/testharnessreport.js"></script>

<script>
"use strict";

test(() => {
  assert_true("HTMLElement" in window, "for-in");
}, "test1: window object should have HTMLElement property in for-in loop");

test(() => {
  assert_true(Object.keys(window).includes("HTMLElement"), "Object.keys()");
}, "test2: window object should have HTMLElement property in Object.keys()");

test(() => {
  assert_equals(Object.getOwnPropertyNames(window).includes("HTMLElement"), "Object.getOwnPropertyNames()");
}, "test3: window object should have HTMLElement property in Object.getOwnPropertyNames()");
</script>
```

**Comment by legendecas** [comment]
`HTMLElement` should be an unenumerable property on `window`, so test2 should expect false?

For test3, it fails on either Node.js versions (v20.17.0, v22.4.1, v22.5.0). So I think it is not relevant to this patch?

**Comment by asamuzaK** [comment]
I see.
Sorry for the noise.

**Comment by legendecas** [comment]
No worries, I will try to address it in a follow-up and see if it will break more tests...

**Review by domenic — APPROVED** [review]

**Pull request merged**